core/client/fs/src/main/java/alluxio/client/block/stream/BufferCachingGrpcDataReader.java

```diff
@@ -0,0 +1,253 @@
+
+import alluxio.client.file.FileSystemContext;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.grpc.ReadRequest;
+import alluxio.grpc.ReadResponse;
+import alluxio.grpc.ReadResponseMarshaller;
+import alluxio.network.protocol.databuffer.DataBuffer;
+import alluxio.network.protocol.databuffer.NioDataBuffer;
+import alluxio.resource.CloseableResource;
+import alluxio.wire.WorkerNetAddress;
+import alluxio.resource.LockResource;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import com.google.common.io.Closer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A gRPC data reader that responsible for reading from a specific block.
+ *
+ * The current implementation cached the block data from the beginning to
+ * the largest index being read.
+ */
+@NotThreadSafe
+public class BufferCachingGrpcDataReader {
+  private static final Logger LOG = LoggerFactory.getLogger(BufferCachingGrpcDataReader.class);
+
+  private final WorkerNetAddress mAddress;
+  private final CloseableResource<BlockWorkerClient> mClient;
+  private final long mDataTimeoutMs;
+  private final ReadRequest mReadRequest;
+  private final GrpcBlockingStream<ReadRequest, ReadResponse> mStream;
+  private final ReadResponseMarshaller mMarshaller;
+  /**
+   * Count the number of threads that are accessing the same block together.
+   * When no thread is accessing this block, the cached data will be GCed.
+   */
+  private final AtomicInteger mRefCount = new AtomicInteger(0);
+  private final Closer mCloser;
+
+  private AtomicInteger mBufferCount = new AtomicInteger(0);
+  private final DataBuffer[] mDataBuffers;
+  private final ReentrantReadWriteLock mBufferLocks = new ReentrantReadWriteLock();
+
+  /** The next pos to read. */
+  private long mPosToRead;
+
+  /**
+   * Creates an instance of {@link BufferCachingGrpcDataReader}.
+   *
+   * @param address the data server address
+   * @param client the block worker client to read data from
+   * @param dataTimeoutMs the maximum time to wait for a data response
+   * @param readRequest the read request
+   * @param stream the underlying gRPC stream to read data
+   */
+  private BufferCachingGrpcDataReader(WorkerNetAddress address,
+      CloseableResource<BlockWorkerClient> client, long dataTimeoutMs,
+      ReadRequest readRequest, GrpcBlockingStream<ReadRequest, ReadResponse> stream) {
+    mAddress = address;
+    mClient = client;
+    mDataTimeoutMs = dataTimeoutMs;
+    mMarshaller = new ReadResponseMarshaller();
+    mPosToRead = readRequest.getOffset();
+    mReadRequest = readRequest;
+    mStream = stream;
+
+    mCloser = Closer.create();
+    mCloser.register(mClient);
```

Comment:
I guess the original way 
```
  @Override
  public void close() throws IOException {
    try {
      if (mClient.get().isShutdown()) {
        return;
      }
      mStream.close();
      mStream.waitForComplete(mDataTimeoutMs);
    } finally {
      mMarshaller.close();
      mClient.close();
    }
  }
```
is used instead of the closer to guarantee the close sequence. The mStream is always close before mClient. So if mClient is shutdown, we know that stream is closed as well. 
in that case, may be better to keep the original way of closing items.

Reply:
Changed back to the original close way to guarantee the order